include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -50,6 +51,246 @@ namespace gko {
 namespace solver {
 
 
+/**
+ * Give a hint about the input of the apply method.
+ */
+enum class input_hint {
+    /**
+     * the input is zero
+     */
+    zero,
+    /**
+     * the input is right hand side
+     */
+    rhs,
+    /**
+     * the input is given
+     */
+    given
+};
+
+
+namespace {
+
+
+inline void fill_zero(LinOp* input)
+{
+    if (auto dense = dynamic_cast<matrix::Dense<float>*>(input)) {
+        dense->fill(gko::zero<float>());
+    } else if (auto dense = dynamic_cast<matrix::Dense<double>*>(input)) {
+        dense->fill(gko::zero<double>());
+    } else if (auto dense =
+                   dynamic_cast<matrix::Dense<std::complex<float>>*>(input)) {
+        dense->fill(gko::zero<std::complex<float>>());
+    } else if (auto dense =
+                   dynamic_cast<matrix::Dense<std::complex<double>>*>(input)) {
+        dense->fill(gko::zero<std::complex<double>>());
+    } else {
+        GKO_NOT_IMPLEMENTED;
+    }
+}
+
+
+}  // namespace
+
+
+class MultigridState;
+
+
+/**
+ * ApplyHint provides a way to give the input hint for apply function.
+ * All functionalities are under protected. It should only be used internally.
+ */
+class ApplyHint {
+protected:
+    friend class MultigridState;
+
+    /**
+     * Applies a linear operator to a vector (or a sequence of vectors) with
+     * a hint.
+     *
+     * Performs the operation x = op(b) with a hint, where op is this linear
+     * operator.
+     *
+     * @param b  the input vector(s) on which the operator is applied
+     * @param x  the output vector(s) where the result is stored
+     * @param hint  the input hint to handle the input vector(s)
+     *
+     * @return this
+     */
+    ApplyHint* apply_with_hint(const LinOp* b, LinOp* x, input_hint hint)
```

Comment:
apply will take the factory parameter as default.
apply_with_hint overrides the original behavior

Reply:
But the `apply_impl` pretty much does exactly what `apply_with_hint` does, just that the first is `virtual`, and the second one a regular function that is overshadowed in the `EnableApplyHint`.
Since both are protected, I just don't see a point in having both.

Reply:
do you mean the apply_impl virtual here and then apply_with_hint is only shown in EnableApplyHint?